File: java/sort_algorithm/impl/MergeSort.java
package sort_algorithm.impl;

import sort_algorithm.Sort;

import java.util.Arrays;

public class MergeSort implements Sort {
    @Override
    public void sort(int[] arr) {
        int[] res = help(arr);
        System.out.println(Arrays.toString(res));
    }

    public int[] help(int[] arr) {
        if (arr.length < 2) return arr;
        int mid = arr.length >> 1;
        int[] l = help(Arrays.copyOfRange(arr, 0, mid)), r = help(Arrays.copyOfRange(arr, mid, arr.length));
        int[] res = new int[arr.length];
        for (int i = 0, j = 0, index = 0;i < l.length || j < r.length;) {
            int rv = j < r.length ? r[j] : Integer.MAX_VALUE, lv = i < l.length ? l[i] : Integer.MAX_VALUE;
            if (rv < lv) {
                res[index++] = rv;
                j++;
            } else {
                res[index++] = lv;
                i++;
            }
        }

        return res;
    }
}

File: golang/sort_algorithm/merge_sort.go
package sort_algorithm

import (
	"fmt"
	"math"
)

type MergeSort []int

func (m *MergeSort) SortImpl() {
	*m = help(*m)
	fmt.Println(m)
}

func help(arr MergeSort) MergeSort {
	if len(arr) < 2 {
		return arr
	}
	mid := len(arr) / 2
	r := help(arr[:mid])
	l := help(arr[mid:])
	res := MergeSort{}
	for i, j := 0, 0; i < len(r) || j < len(l); {
		vi, vj := math.MaxInt32, math.MaxInt32
		if i < len(r) {
			vi = r[i]
		}
		if j < len(l) {
			vj = l[j]
		}
		if vj > vi {
			res = append(res, vi)
			i++
		} else {
			res = append(res, vj)
			j++
		}
	}

	return res
}
